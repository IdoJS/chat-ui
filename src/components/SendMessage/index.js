import React from 'react';
import PropTypes from 'prop-types';

const SendMessage = () => {
  return (
    <div className='send_box'>
      <input className='primary_input' name='inputMessage' type='input'/>
      <button className='secondary_button'>
        send
      </button>
    </div>
  )
};

export default SendMessage;