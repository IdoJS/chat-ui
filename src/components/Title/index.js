import React from 'react';
import PropTypes from 'prop-types';

const Title = (props) => {
  return <h1 className="secondary_title">Hello {props.userName}</h1>;
};

Title.propTypes = {
  userName : PropTypes.string.isRequired
};

Title.defaultProps = {
  userName : ''
};

export default Title