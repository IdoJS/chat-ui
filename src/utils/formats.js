const REGEX_SPLIT_DATE = /[T.]+/;
const LOCAL_OFFSET = new Date().getTimezoneOffset() * 60000;

/**
 * Return split timestamp by timezone
 * [0] - date
 * [1] - time
 * @param timestamp
 */
export const splitTimestamp = (timestamp) => {
  const localTime = timestamp - LOCAL_OFFSET;
  let formattedDateArr;

  try {
    formattedDateArr = new Date(localTime).toISOString().split(REGEX_SPLIT_DATE);
  } catch(e){
    formattedDateArr = ['','',''];
  }
  return formattedDateArr;
};