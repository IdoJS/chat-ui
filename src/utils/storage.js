/**
 * Choose storage by ENV
 * @type {Storage}
 */
const storage = !process.env.NODE_ENV || process.env.NODE_ENV === 'development' ? sessionStorage : localStorage;

/**
 * Generate unique ID
 * https://stackoverflow.com/questions/105034/create-guid-uuid-in-javascript
 * @returns {string}
 */
const guid = () => {
  const s4 = () => {
    return Math.floor((1 + Math.random()) * 0x10000)
      .toString(16)
      .substring(1);
  };
  return s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() + s4() + s4();
};

const STORAGE_AVATAR_CLASS = 'avatarClass';
const STORAGE_USER_NAME = 'userName';
const STORAGE_USER_ID = 'UserID';

export const getAvatarClass = () => {
  return storage.getItem(STORAGE_AVATAR_CLASS);
};

export const getUserName = () => {
  return storage.getItem(STORAGE_USER_NAME);
};

export const setAvatarClass = (avatar) => {
  storage.setItem(STORAGE_AVATAR_CLASS, avatar);
};

export const setUserName = (userName) => {
  storage.setItem(STORAGE_USER_NAME, userName);
};

export const getUserId = () => {
  return storage.getItem(STORAGE_USER_ID);
};

export const setUserId = () => {
  storage.setItem(STORAGE_USER_ID, guid());
};