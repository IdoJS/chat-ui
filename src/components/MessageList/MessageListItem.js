import React from 'react';
import PropTypes from 'prop-types';
import {splitTimestamp} from '../../utils/formats';

const MessageListItem = (props) => {
  const {data, currentUserId} = props;
  const {text, userName, timestamp, avatar, userId} = data;
  const currentTimeArr = splitTimestamp(timestamp);
  const isCurrentUserMsg = userId === currentUserId;

  return <li className='chat_item'>
    <div className={`chat_item_text_header chat_item${isCurrentUserMsg ? '_user' : '_default'}`}>
      <h3 className='chat_item_user_name'>{userName}</h3>
      <h3>{currentTimeArr[1]}</h3>
    </div>
    <div className={`ui message chat_item_msg chat_item_msg${isCurrentUserMsg ? '_user' : '_default'}`}>
      <div></div>
      <div className={`chat_item_img avatar ${avatar}`}/>
      <div className={`chat_item_text chat_item_msg chat_item_msg${isCurrentUserMsg ? '_user' : '_default'}`}>
        <p
          className={`ui ${isCurrentUserMsg ? 'left pointing blue' : 'right pointing teal'} chat_item_msg${isCurrentUserMsg ? '_user' : '_default'} basic label`}>{text}</p>
      </div>
    </div>
  </li>
};

MessageListItem.propTypes = {
  currentUser: PropTypes.string.isRequired,
  currentUserId: PropTypes.string.isRequired,
  data: PropTypes.shape({
    text: PropTypes.string.isRequired,
    userName: PropTypes.string.isRequired,
    timestamp: PropTypes.number.isRequired
  }).isRequired
};

MessageListItem.defaultProps = {
  currentUser: '',
  data: {
    text: '',
    userName: '',
    timestamp: Date.now()
  }
};

export default MessageListItem;


