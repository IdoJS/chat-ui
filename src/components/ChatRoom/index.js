import React from 'react';
import PropTypes from 'prop-types';
import socketObserver from '../../utils/socketObserver';

import RequestTypes from '../../requestTypes';

import Title from '../Title';
import SendMessage from '../SendMessage';
import MessageList from '../MessageList';

import {getUserName, getAvatarClass} from '../../utils/storage';

class ChatRoom extends React.Component {

  constructor(props) {
    super(props);

    this.currentUserName = getUserName();
    this.currentAvatarClass = getAvatarClass();

    this.onMessageSend = this.onMessageSend.bind(this);
    this.onMessageTyping = this.onMessageTyping.bind(this);


  }

  onMessageSend({text}) {
    const data = {
      text,
      userName: this.currentUserName,
      timestamp: Date.now(),
      avatar: this.currentAvatarClass,
      isTyping: false,
      requestType: RequestTypes.REQUEST_SENDING
    };

    socketObserver.send(data);
    this.props.userSendMessage && this.props.userSendMessage(data);
  }

  onMessageTyping(isTyping) {
    socketObserver.send({
      text: '...',
      userName: this.currentUserName,
      timestamp: Date.now(),
      avatar: this.currentAvatarClass,
      isTyping,
      requestType: RequestTypes.REQUEST_TYPING
    });
  }

  componentDidMount() {
    socketObserver.subscribe({
      identifier: this,
      o: function (data) {
        if (data.requestType === RequestTypes.REQUEST_TYPING) {
          if(data.isTyping){
            this.props.updateMessageTypingAdd && this.props.updateMessageTypingAdd(data)
          } else {
            this.props.updateMessageTypingRemove && this.props.updateMessageTypingRemove(data)
          }
        } else {
          this.props.updateMessageList(data);
        }
      }.bind(this)
    });
  }

  componentWillUnmount() {
    socketObserver.unSubscribe({
      identifier: this
    });
  }

  shouldComponentUpdate(nextProps, nextState) {
    const {chatData} = this.props;

    return chatData.typingArr.length !== nextProps.chatData.typingArr.length ||
      chatData.chatArr.length !== nextProps.chatData.chatArr.length ||
      chatData.loading !== nextProps.chatData.loading
  }

  render() {
    return (<div className='chat'>
      <Title userName={this.currentUserName}
             currentUser={this.currentUserName}/>

      <MessageList chatData={this.props.chatData}
                   currentUser={this.currentUserName}/>

      <SendMessage isLoading={this.props.chatData.loading}
                   onMessageSend={this.onMessageSend}
                   onMessageTyping={this.onMessageTyping}/>
    </div>);
  }
};


ChatRoom.defaultProps = {
  chatData: {
    loading: false,
    timestamp: Date.now(),
    chatArr: [],
    typingArr: [],
    timeout: false
  },
  updateMessageList: () => {
    console.warn('Please implement updateMessageList');
  },
  updateMessageTypingAdd: () => {
    console.warn('Please implement updateMessageTypingAdd');
  },
  updateMessageTypingRemove: () => {
    console.warn('Please implement updateMessageTypingRemove');
  }
};

ChatRoom.propTypes = {
  chatData: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
    chatArr: PropTypes.array.isRequired,
    typingArr: PropTypes.array.isRequired,
    timeout: PropTypes.bool.isRequired
  }).isRequired,
  updateMessageList: PropTypes.func.isRequired,
  updateMessageTypingAdd : PropTypes.func.isRequired,
  updateMessageTypingRemove : PropTypes.func.isRequired
};

export default ChatRoom;