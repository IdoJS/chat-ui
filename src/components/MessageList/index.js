import React from 'react';
import MessageListItem from './MessageListItem';
import PropTypes from 'prop-types';

class MessageList extends React.PureComponent {
  render() {
    const {chatData, currentUser} = this.props;
    const {chatArr} = chatData;
    const items = chatArr.map((data, key) => {
      return <MessageListItem {...data} currentUser={currentUser} key={key}/>
    });

    return <ul className='chat_box' id='chat_container'>{items}</ul>
  }

  componentDidUpdate() {
    const objDiv = document.getElementById('chat_container');
    objDiv.scrollTop = objDiv.scrollHeight;
  }
}

MessageList.propTypes = {
  currentUser: PropTypes.string.isRequired,
  chatData: PropTypes.shape({
    chatArr: PropTypes.array.isRequired
  }).isRequired
};

export default MessageList;