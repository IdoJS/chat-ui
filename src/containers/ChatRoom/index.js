import React from 'react';
import {connect} from 'react-redux';

import ChatRoom from '../../components/ChatRoom';
import {messageSend, messageReceived, messageTyping} from '../../actions/chatAction';

const mapDispatchToProps = dispatch => {
  return {
    userSendMessage: (data) => dispatch(messageSend(data)),
    updateMessageList: (data) => dispatch(messageReceived(data)),
    updateMessageTyping: (data) => dispatch(messageTyping(data))
  };
};

const mapStateToProps = state => {
  return {
    chatData: state
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ChatRoom);