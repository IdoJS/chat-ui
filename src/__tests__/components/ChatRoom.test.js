import React from 'react';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';
import {setAvatarClass, setUserName} from '../../utils/storage';

import ChatRoom from '../../../src/components/ChatRoom';
import Title from '../../../src/components/Title';
import SendMessage from '../../../src/components/SendMessage';
import MessageList from '../../../src/components/MessageList';

describe('ChatRoom component', () => {

  beforeAll(()=>{
    setAvatarClass('MockAvatar');
    setUserName('MockUser');
  });

  it('Should render the ChatRoom view', () => {
    const wrapper = shallow(<ChatRoom/>);
    const title = wrapper.find(Title);
    const sendMsg = wrapper.find(SendMessage);
    const msgList = wrapper.find(MessageList);

    expect(title).toHaveLength(1);
    expect(sendMsg).toHaveLength(1);
    expect(msgList).toHaveLength(1);

  });

  it('SnapShot rendering ChatRoom', () => {
    const wrapper = shallow(<ChatRoom/>);
    const tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });

});