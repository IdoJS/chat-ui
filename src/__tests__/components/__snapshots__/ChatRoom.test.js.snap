// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatRoom component SnapShot rendering ChatRoom 1`] = `
<div
  className="chat_box_container"
>
  <h1
    className="secondary_title"
  >
    Hello 
    mock_avatar_class
  </h1>
  <ul
    className="chat_box"
    id="chat_container"
  />
  <div
    className="send_box"
  >
    <input
      className="primary_input"
      disabled={false}
      name="inputMessage"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Type message..."
      type="input"
      value=""
    />
    <button
      className="secondary_button"
      disabled={true}
      onClick={[Function]}
    >
      send
    </button>
  </div>
</div>
`;
